### ISSUE TITLE
Dispatch error events when event handler attributes fail to compile.

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#13152]
I'm not currently doing that because of double borrows in `html/webappapis/scripting/processing-model-2/runtime-error-in-body-onerror.html`.

We should also make sure to pass `tests/wpt/web-platform-tests/html/webappapis/scripting/events/invalid-uncompiled-raw-handler-compiled-late.html`; I believe that we'll need to change the timing of compilation for that.

**Comment by Ms2ger** [comment]
Should also write tests that add/remove event listeners from the error event handler.

**Comment by Ms2ger** [comment]
> when compiling the event handlers, we borrow eventtarget.handlers, and then we call back into script with that borrow held

**Comment by jdm** [comment]
@ofekd Does this look interesting? This is in the code in `components/script/dom/eventtarget.rs`.

**Comment by ofekd** [comment]
@jdm Yeah I can do this one. Though take into account I'm still not fast solving things in Servo

**Comment by jdm** [comment]
That's fine! Please ask questions if anything's unclear.

**Comment by ofekd** [comment]
@Ms2ger I've spent some time learning the events code, but I still miss the double borrows, and how it's all specific to `body onerror`.

**Comment by jdm** [comment]
@ofekd I believe the problem is that if we dispatch an error event when compilation fails, it triggers a double borrow error. Because of this, we do not dispatch an error event right now. I haven't verified this, however.

**Comment by Ms2ger** [comment]
Say we have
```html
<body onerror="f()" onclick="}">
<script>
function f() { ... }
document.body.dispatchEvent(new Event("click"));
</script>
```

When `dispatchEvent` is called, we will borrow the body element's `handlers` field (mutably), try to compile the `onclick` attribute's value, and fail. Then we'll try to dispatch an `error` event, which will try to borrow the same `handlers` field, and this panics.

I don't have a clear idea of how to fix this, but I do think the specification requires us to compile the just before calling the event handler. It is probably possible to detect with this test case:

```html
<body onerror="f()">
<script>
function f() { ... }
var body = document.body;
body.addEventListener("click", function g() { ... });
body.setAttribute("onclick", "}");
body.dispatchEvent(new Event("click"));
</script>
```

In this case, it is my understanding `g()` should be called before `f()`.

**Comment by ofekd** [comment]
Hi all

I'd like to apologize --

I took this when I was relatively available and then literally days later got held up pretty bad with work. This is not what I meant when I said I'm not that fast with Servo.

Since pressure is soon off I can be right at it again, if that's fine by you.

**Comment by jdm** [comment]
That's fine! Thanks for letting us know!

**Comment by nox** [comment]
`/html/webappapis/scripting/events/invalid-uncompiled-raw-handler-compiled-late.html` still fails.

**Comment by pshaughn** [comment]
The first part of https://github.com/servo/servo/issues/13152#issuecomment-254849922 might describe a problem that no longer exists.

When I change https://github.com/servo/servo/blob/e10cf7adad91a4740a5ba859f4fc33ee59423908/components/script /dom/eventtarget.rs#L532 from false to true and go to a page with the following source:

```
<body onerror="f()" onclick="}">
  <div id="output">Script not started.</div>
  <script>
    function f() {
     document.getElementById("output").innerText="called f without crashing!";
    }
    document.body.dispatchEvent(new Event("click"));
  </script>
```
I see "called f without crashing", with no double-borrow panic.

When I try the more complicated second example, f fires before g, so there are other matters to consider, but it may be that the double-borrow hazard for nested events has been eliminated by other code changes and it is safe to start throwing event handler attribute compile errors.

**Comment by pshaughn** [comment]
That PR does not close this because of the still-open timing issue tested by html/webappapis/scripting/events/invalid-uncompiled-raw-handler-compiled-late.window.html.

**Comment by pshaughn** [comment]
if I'm reading https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/event_handler_syntax_error.html right, we shouldn't be compiling strings in random "onwhatever" attributes; this behavior's probably in about the same place as the timing issue and might be solved in the same change.

**Comment by pshaughn** [comment]
https://github.com/servo/servo/blob/master/components/script/dom/htmlelement.rs#L728 seems to be at fault: any attribute starting with "on" gets stored as an uncompiled event handler instead of a string, even when it isn't an event handler attribute.

**Comment by pshaughn** [comment]
Less terrifying than it could be; https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers-on-elements,-document-objects,-and-window-objects lists the handlers and it **looks like** there are only two kinds of match to make since we already know we're in an element: either it has to reflect window handlers, or it doesn't, and those **seem** to be the only relevant differentiators as to whether an on* attribute needs compilation, emphasis deliberate.

**Comment by pshaughn** [comment]
All of the above is written with the expectation that any on* content attribute we're supposed to use for event listening corresponds to an IDL attribute and vice versa; if there's some case where that expectation isn't true, then it gets much messier. If they are all in the IDL, then codegen could be empowered to write an "is this DOMString the name for an event handler attribute on this element" method just by compile-time IDL introspection, plus maybe a runtime flag for whether to consider window reflection.

**Comment by pshaughn** [comment]
I had not realized there were already event-handler macros being maintained separately from their matching IDL attributes; my previous musings on how to track them automatically are unnecessary in light of that.

**Comment by Manishearth** [comment]
Filed https://github.com/servo/servo/issues/25760 for doing the event handler thing the right way

**Comment by CYBAI** [comment]
Just add a note: with fixing this, we can fix the TIMEOUT for the `test3` in `/html/webappapis/scripting/events/compile-event-handler-lexical-scopes.html` 👀
